### PR TITLE
Fantomas: support cursor api

### PR DESCRIPTION
**Pull request by DedSec256**
https://github.com/fsprojects/fantomas/issues/2727

### CODE REVIEW

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeCleanup/FSharpReformatCode.fs [hunk 1]
```diff
@@ -65,23 +68,33 @@ type FSharpReformatCode() =
...
+
+                if isNull newCursorPosition then () else
+
+                // move cursor after current document transaction
+                let moveCursorLifetime = new LifetimeDefinition()
+                let codeCleanupService = solution.GetComponent<CodeCleanupService>()
+                codeCleanupService.WholeFileCleanupCompletedAfterSave.Advise(moveCursorLifetime.Lifetime, fun _ ->
+                    moveCursorLifetime.Terminate()
+                    textControl.Caret.MoveTo(Int32<DocLine>.op_Explicit(newCursorPosition.Row),
+                                             Int32<DocColumn>.op_Explicit(newCursorPosition.Column),
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeCleanup/FSharpReformatCode.fs:99]: Would `docLine` and `docColumn` helpers work here?

File: rider-fsharp/testData/fantomas/FantomasRunOptionsTest/local tool 6_0 with cursor/source/LargeFile.fs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                if isNull newCursorPosition then () else
+
+                // move cursor after current document transaction
+                let moveCursorLifetime = new LifetimeDefinition()
+                let codeCleanupService = solution.GetComponent<CodeCleanupService>()
+                codeCleanupService.WholeFileCleanupCompletedAfterSave.Advise(moveCursorLifetime.Lifetime, fun _ ->
+                    moveCursorLifetime.Terminate()
+                    textControl.Caret.MoveTo(Int32<DocLine>.op_Explicit(newCursorPosition.Row),
+                                             Int32<DocColumn>.op_Explicit(newCursorPosition.Column),
+                                             CaretVisualPlacement.Generic))
```
Comment (auduchinok) [inline rider-fsharp/testData/fantomas/FantomasRunOptionsTest/local tool 6_0 with cursor/source/LargeFile.fs:98]: Can a simpler test data be used here?

Comment (DedSec256) [reply rider-fsharp/testData/fantomas/FantomasRunOptionsTest/local tool 6_0 with cursor/source/LargeFile.fs:98]: it's not easy enough :D

File: ReSharper.FSharp/src/FSharp.Fantomas.Host/src/FantomasCodeFormatter.cs [hunk 3]
```diff
@@ -152,35 +158,64 @@ public static string FormatSelection(RdFantomasFormatSelectionArgs args)
...
+    public static RdFormatResult FormatDocument(RdFantomasFormatDocumentArgs args)
+    {
+      var formatDocumentOptions = GetFormatDocumentOptions(args);
+      var formatDocumentAsync = CodeFormatterType.InvokeMember("FormatDocumentAsync",
+        BindingFlags.Static | BindingFlags.InvokeMethod | BindingFlags.Public, Type.DefaultBinder, null,
+        formatDocumentOptions);
+      var formatResult = FSharpAsync.StartAsTask((dynamic)formatDocumentAsync, null, null).Result;
+
+      if (CurrentVersion < Version60)
+        return new RdFormatResult(formatResult.Replace("\r\n", args.NewLineText), null);
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Fantomas.Host/src/FantomasCodeFormatter.cs:193]: Could you add a comment why this is needed, please?

Comment (nojaf) [reply ReSharper.FSharp/src/FSharp.Fantomas.Host/src/FantomasCodeFormatter.cs:193]: You should be able to pass in `lf` as part of the [configuration](https://fsprojects.github.io/fantomas/docs/end-users/Configuration.html#end_of_line) btw.
(And override whatever the user had)

Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Fantomas.Host/src/FantomasCodeFormatter.cs:193]: This is a relic of the past. We can try to rewrite it in a separate PR.
https://github.com/JetBrains/resharper-fsharp/blob/80620e2c051ab1a9d30367565ef60f760cdc1639/ReSharper.FSharp/src/FSharp.Psi.Features/src/Formatter/FSharpReformatCode.fs#L98

Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Fantomas.Host/src/FantomasCodeFormatter.cs:193]: I think a comment about the version check would stil be good here. With a link to a PR or a commit.

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeCleanup/FSharpReformatCode.fs [hunk 4]
```diff
@@ -65,23 +67,33 @@ type FSharpReformatCode() =
...
+                    let formatted = fantomasHost.FormatSelection(filePath, range, text, settings, parsingOptions, newLineText)
+                    let offset = rangeMarker.DocumentRange.StartOffset.Offset
+                    let oldLength = rangeMarker.DocumentRange.Length
+                    let documentChange = DocumentChange(document, offset, oldLength, formatted, stamp, modificationSide)
+                    use _ = WriteLockCookie.Create()
+                    document.ChangeDocument(documentChange, TimeStamp.NextValue)
+                    sourceFile.GetPsiServices().Files.CommitAllDocuments()
+                with _ -> ()
+            else
+                let textControl = textControlManager.VisibleTextControls |> Seq.find (fun c -> c.Document == document)
```
[the remark this answers is not in the document]
Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeCleanup/FSharpReformatCode.fs:82]: Thanks for writing about it, but are there any other details? 🙂
Where exactly, what exception type, etc.

[the remark this answers is not in the document]
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeCleanup/FSharpReformatCode.fs]: Yes, sure. In the case of a project reformat and cleanup, for example, the document being formatted may not be open, and therefore will not be found in the list of visible text controls

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeCleanup/FSharpReformatCode.fs [hunk 5]
```diff
@@ -79,8 +79,8 @@ type FSharpReformatCode(textControlManager: ITextControlManager) =
                     sourceFile.GetPsiServices().Files.CommitAllDocuments()
                 with _ -> ()
             else
-                let textControl = textControlManager.VisibleTextControls |> Seq.find (fun c -> c.Document == document)
-                let cursorPosition = textControl.Caret.Position.Value.ToDocLineColumn()
+                let textControl = textControlManager.VisibleTextControls |> Seq.tryFind (fun c -> c.Document == document)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeCleanup/FSharpReformatCode.fs:82]: What if there are multiple editors with the same document? This check seems to ignore which editor was actually focused before the reformat. Please try to use data contexts or other ways to get the proper text control.

**Review by auduchinok — APPROVED** [review]

**Pull request merged**